extras/kms/repository_root_key_version.go

```diff
@@ -174,3 +176,96 @@ func (r *repository) ListRootKeyVersions(ctx context.Context, keyWrapper wrappin
+// This function encapsulates all the work required within a dbw.TxHandler and
+// allows this capability to be shared with other repositories or just called
+// within a transaction.  To be clear, this repository function doesn't include
+// its own transaction and is intended to be used within a transaction provide
+// by the caller.
+func rewrapRootKeyVersionsTx(ctx context.Context, reader dbw.Reader, writer dbw.Writer, rootWrapper wrapping.Wrapper, rootKeyId string, _ ...Option) error {
+	const (
+		op           = "kms.rewrapRootKeyVersionsTx"
+		keyFieldName = "CtKey"
+	)
+	if isNil(reader) {
+		return fmt.Errorf("%s: missing reader: %w", op, ErrInvalidParameter)
+	}
+	if isNil(writer) {
+		return fmt.Errorf("%s: missing writer: %w", op, ErrInvalidParameter)
+	}
+	if isNil(rootWrapper) {
+		return fmt.Errorf("%s: missing root wrapper: %w", op, ErrInvalidParameter)
+	}
+	if rootKeyId == "" {
+		return fmt.Errorf("%s: missing root key id: %w", op, ErrInvalidParameter)
+	}
+	r, err := newRepository(reader, writer)
+	if err != nil {
+		return fmt.Errorf("%s: unable to create repo: %w", op, err)
+	}
+	// rewrap the rootKey versions using the scope's root key to find them
+	rkvs, err := r.ListRootKeyVersions(ctx, rootWrapper, rootKeyId, withReader(reader))
+	if err != nil {
+		return fmt.Errorf("%s: unable to list root key versions: %w", op, err)
+	}
+	for _, kv := range rkvs {
+		if err := kv.Encrypt(ctx, rootWrapper); err != nil {
+			return fmt.Errorf("%s: failed to rewrap root key version: %w", op, err)
+		}
+		rowsAffected, err := writer.Update(ctx, kv, []string{keyFieldName}, nil, dbw.WithVersion(&kv.Version))
+		if err != nil {
+			return fmt.Errorf("%s: failed to update root key version: %w", op, err)
+		}
+		if rowsAffected != 1 {
+			return fmt.Errorf("%s: expected to update 1 root key version and updated %d", op, rowsAffected)
+		}
+	}
+	return nil
+}
+
+// rotateRootKeyVersionTx will rotate the key version for the given rootKeyId.
+// This function encapsulates all the work required within a dbw.TxHandler and
+// allows this capability to be shared with other repositories or just called
+// within a transaction.  To be clear, this repository function doesn't include
+// its own transaction and is intended to be used within a transaction provide
+// by the caller.
+// Supported options: withRandomReader
+func rotateRootKeyVersionTx(ctx context.Context, writer dbw.Writer, rootWrapper wrapping.Wrapper, rootKeyId string, opt ...Option) (*rootKeyVersion, error) {
+	const op = "kms.rotateRootKeyVersionTx"
+	if isNil(rootWrapper) {
+		return nil, fmt.Errorf("%s: missing root wrapper: %w", op, ErrInvalidParameter)
+	}
+	if rootKeyId == "" {
+		return nil, fmt.Errorf("%s: missing root key id: %w", op, ErrInvalidParameter)
+	}
+
+	if isNil(writer) {
+		return nil, fmt.Errorf("%s: missing writer: %w", op, ErrInvalidParameter)
+	}
+	opts := getOpts(opt...)
+	if isNil(opts.withRandomReader) {
+		opts.withRandomReader = rand.Reader
```

Comment:
This seems like it should be part of defaultOpts?

Reply:
agreed.  ty!